### PR TITLE
Apply "qualified do" notation to bangs and comprehensions

### PR DESCRIPTION
**Pull request by Russoul**
Example in the test file is kind of silly and meaningless though.

Comprehensions might not be the best fit for qualified do, because they target `Applicative`'s and `Alternative`'s.
However, it probably won't hurt to include them too.
As a possible future extension, we could also qualify user-written `pure` and `guard` calls as well.

### CODE REVIEW

**Comment by edwinb** [comment]
Good idea, thanks. Please could you update the tests to resolve the conflicts, though? Then I'll merge - this might be a victim of my recent merge frenzy, sorry about that...

**Comment by Russoul** [comment]
@edwinb All done

File: src/Idris/Desugar.idr [hunk 1]
```diff
@@ -392,15 +393,21 @@ mutual
...
+  desugarB side ps (PComprehension fc ret conds) = do
+        let ns = mbNamespace !(get Bang)
+        desugarB side ps (PDoBlock fc ns (map (guard ns) conds ++ [toPure ns ret]))
     where
-      guard : PDo -> PDo
-      guard (DoExp fc tm) = DoExp fc (PApp fc (PRef fc (UN "guard")) tm)
-      guard d = d
+      mbNS : Maybe Namespace -> Name -> Name
+      mbNS Nothing n = n
+      mbNS (Just ns) n = NS ns n
```
Comment (gallais) [inline src/Idris/Desugar.idr:402]: Can we please lift this to `Core.Name`?

**Comment by stefan-hoeck** [comment]
Just a quick question: Does this also work for idiom brackets (`[| a + b |]`). Would be great to have that as well.

**Comment by Russoul** [comment]
@stefan-hoeck I haven't tested. They are desugared into `pure`, among other things. `pure` is not covered by the PR, that's for sure

**Comment by edwinb** [comment]
Idiom brackets are a different bit of desugaring that don't necessarily come under do blocks, so it'd need a different approach unfortunately. The qualified do notation was really just about solving the occasional problem of having to be explicit about which `>>=` you need if you have some fancy DSL.

**Comment by Russoul** [comment]
Personally, I use qualified do for handling effects (as composition of monads), by specialising (>>=) to an instance of a wanted composition in a dedicated namespace.

**Pull request merged**